File: frontend/src/types/domain.ts
/** Domain contracts mirrored from backend Pydantic schemas. No business logic here. */

export type SensorType = "sentinel-2" | "sentinel-1";
export type DataMode = "development" | "earth_engine";
export type AnalysisStatus = "pending" | "running" | "completed" | "failed";
export type TraceStatus = "pending" | "running" | "completed" | "failed";

export interface GeoJSONGeometry {
  type: "Polygon" | "MultiPolygon" | "Point" | "LineString";
  coordinates: number[][][] | number[][] | number[];
}

export interface AOI {
  geometry: GeoJSONGeometry;
  name?: string | null;
  area_km2?: number | null;
}

export interface ImageryPreferences {
  cloud_cover_max?: number;
  prefer_least_cloud?: boolean;
}

export interface ImageryRequest {
  aoi: AOI;
  start_date: string;
  end_date: string;
  sensor?: SensorType;
  preferences?: ImageryPreferences;
}

export interface ImageryScene {
  scene_id: string;
  acquisition_date: string;
  cloud_cover_percent?: number | null;
  preview_url?: string | null;
}

export interface SpatialMetadata {
  crs?: string;
  bbox: number[];
  resolution_m?: number | null;
}

export interface ImageryResult {
  source: string;
  mode: DataMode;
  sensor: SensorType;
  scenes: ImageryScene[];
  spatial: SpatialMetadata;
  message?: string | null;
}

export interface QueryRequest {
  query: string;
  aoi?: AOI;
  earlier_date?: string;
  later_date?: string;
  sensor?: SensorType;
  preferences?: ImageryPreferences;
  image_id?: string;
  earlier_image_id?: string;
  later_image_id?: string;
  optical_image_id?: string;
  sar_image_id?: string;
  demo_mode?: boolean;
}

export type ImageModality = "optical" | "multispectral" | "sar";
export type ImageFormat = "geotiff" | "tiff" | "png" | "jpeg";

export interface ImageInput {
  id: string;
  modality: ImageModality;
  format: ImageFormat;
  filename: string;
  width: number;
  height: number;
  georeferenced: boolean;
  acquisition_datetime?: string | null;
  benchmark_dataset?: boolean;
  co_registered_benchmark?: boolean;
  benchmark_pair_id?: string | null;
  bounds?: number[] | null;
  crs?: string | null;
  transform?: number[] | null;
  native_crs?: string | null;
  native_bounds?: number[] | null;
}

export interface UploadImageResponse {
  image: ImageInput;
}

export interface SingleImageVQAResult {
  task: "single_image_vqa";
  answer: string;
  model_name: string;
  model_version: string;
  provider: "development" | "geochat_service";
  provenance: string;
  confidence?: number | null;
  confidence_available: boolean;
  input_image_id: string;
  requested_modality: string;
  inference_metadata?: Record<string, unknown>;
}

export interface SingleImageCaptionResult {
  task: "single_image_caption";
  description: string;
  model_name: string;
  model_version: string;
  provider: "development" | "geochat_service";
  provenance: string;
  confidence?: number | null;
  confidence_available: boolean;
  input_image_id: string;
  requested_modality: string;
  inference_metadata?: Record<string, unknown>;
}

export interface CrossModalOpticalSARResult {
  task: "cross_modal_optical_sar";
  answer: string;
  question: string;
  optical_analysis: {
    modality: string;
    summary: string;
    analyzer: string;
    provider: string;
    confidence_available: boolean;
  };
  sar_analysis: {
    modality: string;
    summary: string;
    analyzer: string;
    provider: string;
    confidence_available: boolean;
  };
  fused_analysis: {
    summary: string;
    fusion_policy: string;
    fused_region_count: number;
    complementary_notes: string[];
  };
  co_registration_status: string;
  co_registration_provenance: string;
  optical_image_id: string;
  sar_image_id: string;
  provider: string;
  provenance: string;
  confidence?: number | null;
  confidence_available: boolean;
}

export interface BiTemporalSceneMetrics {
  changed_pixel_count?: number | null;
  total_pixel_count?: number | null;
  changed_percentage?: number | null;
  area_m2?: number | null;
  area_ha?: number | null;
  area_km2?: number | null;
  region_count?: number | null;
}

export interface BiTemporalDetectorSummary {
  detector: string;
  algorithm?: string | null;
  detector_version?: string | null;
  primary_index?: string | null;
  change_direction_hint?: string | null;
  histogram_confidence?: number | null;
  confidence_kind?: "histogram_separability" | null;
}

export interface BiTemporalImageProvenance {
  earlier_source_ref?: string | null;
  later_source_ref?: string | null;
  earlier_filename?: string | null;
  later_filename?: string | null;
  earlier_modality?: string | null;
  later_modality?: string | null;
  earlier_band_names?: string[] | null;
  later_band_names?: string[] | null;
  crs?: string | null;
  coregistration_performed?: boolean | null;
  positional_band_fallback_used?: boolean | null;
}

export interface BiTemporalChangeResult {
  task: "bi_temporal_change_vqa";
  change_summary: string;
  question: string;
  changed_region_count: number;
  change_map_available: boolean;
  detector: string;
  provider: "development" | "uploaded_cva";
  provenance: string;
  confidence?: number | null;
  confidence_available: boolean;
  confidence_kind?: "histogram_separability" | null;
  earlier_image_id: string;
  later_image_id: string;
  earlier_acquisition: string;
  later_acquisition: string;
  earlier_date: string;
  later_date: string;
  scene_metrics?: BiTemporalSceneMetrics | null;
  detector_summary?: BiTemporalDetectorSummary | null;
  image_provenance?: BiTemporalImageProvenance | null;
  inference_metadata?: Record<string, unknown>;
}

export interface BiTemporalRegionInterpretationResult {
  task: "bi_temporal_region_interpretation";
  answer: string;
  region_id: string;
  session_id: string;
  question: string;
  detector: string;
  region_confidence: number;
  confidence_kind?: "histogram_separability" | null;
  change_direction_hint?: string | null;
  model_name: string;
  model_version: string;
  provider: "development" | "geochat_service";
  provenance: string;
  confidence_available: boolean;
  inference_metadata?: Record<string, unknown>;
  earlier_image_id: string;
  later_image_id: string;
  preview_bbox_wgs84: string;
  evidence_inputs: "before_after_composite_crop";
}

export interface InterpretRegionData {
  interpretation: BiTemporalRegionInterpretationResult;
  trace_step: TraceStep;
}

export interface ConversationTurnRecord {
  turn_id: string;
  turn_index: number;
  user_message: string;
  assistant_answer: string;
  created_at: string;
  route?: "geo" | "general" | null;
  provider?: "development" | "geochat_service" | "groq" | null;
  scope?: "selected_region" | "general_assistant" | null;
}

export interface RegionConversationRecord {
  conversation_id: string;
  session_id: string;
  region_id: string;
  turns: ConversationTurnRecord[];
}

export interface BiTemporalRegionChatResult {
  task: "bi_temporal_region_chat";
  answer: string;
  session_id: string;
  region_id: string;
  conversation_id: string;
  turn_id: string;
  turn_index: number;
  message: string;
  detector: string;
  region_confidence: number;
  confidence_kind?: "histogram_separability" | null;
  change_direction_hint?: string | null;
  model_name: string;
  model_version: string;
  provider: "development" | "geochat_service" | "groq";
  provenance: string;
  confidence_available: boolean;
  inference_metadata?: Record<string, unknown>;
  route: "geo" | "general";
  classification: "geo" | "general" | "ambiguous";
  scope: "selected_region" | "general_assistant";
  preview_bbox_wgs84: string;
  evidence_inputs: "before_after_composite_crop" | "general_assistant_no_imagery";
  scope_limited: boolean;
  conversation: RegionConversationRecord;
}

export interface RegionChatData {
  chat: BiTemporalRegionChatResult;
  trace_step: TraceStep;
}

export type GroundSceneCategory =
  | "vegetation_loss"
  | "built_up_increase"
  | "water_shrinkage"
  | "flood"
  | "generic_change";

export interface GroundContextLocation {
  latitude: number;
  longitude: number;
}

export interface GroundContextScene {
  category: GroundSceneCategory;
  title: string;
  description: string;
  features: string[];
}

export interface GroundContextImage {
  asset_id: string;
  url: string;
  alt: string;
}

export interface GroundContextProvenance {
  provider: "mock_ground_context";
  source_type: "mock";
  status: "demonstration_data";
  real_world_imagery: false;
  disclosure: string;
}

export interface GroundContextResult {
  session_id: string;
  region_id: string;
  location: GroundContextLocation;
  heading: number;
  capture_date: string;
  scene: GroundContextScene;
  image: GroundContextImage;
  provenance: GroundContextProvenance;
}

export interface Metric {
  name: string;
  value: number | string;
  unit?: string | null;
  source: string;
}

export interface EvidenceRegion {
  id: string;
  geometry: GeoJSONGeometry;
  type: string;
  confidence: number;
  metrics: Metric[];
  source: string;
  metadata: Record<string, unknown>;
}

export interface PlanQueryMetadata {
  planner?: string;
  intent?: string;
  required_tools?: string[];
  requested_modalities?: string[];
  planner_version?: string;
  fallback_used?: boolean;
  status?: string;
  duration_ms?: number;
}

export interface FetchImageryMetadata {
  imagery_strategy?: string;
  composite_method?: string;
  demonstration_data?: boolean;
  fallback_events?: Array<Record<string, unknown>>;
  fallback_policy?: string;
  t1?: {
    requested_date?: string;
    window_start?: string;
    window_end?: string;
    scene_count?: number;
    scene_dates?: string[];
    fallback?: Record<string, unknown>;
  };
  t2?: {
    requested_date?: string;
    window_start?: string;
    window_end?: string;
    scene_count?: number;
    scene_dates?: string[];
    fallback?: Record<string, unknown>;
  };
}

export interface TraceStep {
  id: string;
  tool_name: string;
  status: TraceStatus;
  started_at?: string | null;
  completed_at?: string | null;
  duration_ms?: number | null;
  summary?: string | null;
  error?: string | null;
  metadata?: PlanQueryMetadata | Record<string, unknown> | null;
}

export type BuildingStatus = "unchanged" | "new" | "removed" | "significantly_changed";

export interface BuildingFootprint {
  id: string;
  geometry: GeoJSONGeometry;
  bbox: [number, number, number, number];
  area_m2: number;
  confidence: number;
  status: BuildingStatus;
  iou_with_match?: number | null;
  matched_id?: string | null;
}

export interface BuildingDetectionResult {
  image_id: string;
  count: number;
  detections: BuildingFootprint[];
  confidence: number;
  detector_name: string;
  total_area_m2: number;
  metrics: Metric[];
}

export interface BuildingTemporalMatchResult {
  earlier_image_id: string;
  later_image_id: string;
  before_count: number;
  after_count: number;
  new_count: number;
  removed_count: number;
  unchanged_count: number;
  changed_count: number;
  confidence: number;
  matcher_name: string;
  matched_footprints: BuildingFootprint[];
  metrics: Metric[];
}

export type SurfaceDomainKind = "built_up" | "water" | "vegetation";

export interface SurfaceAreaChangeResult {
  domain: SurfaceDomainKind;
  earlier_image_id: string;
  later_image_id: string;
  before_area_m2: number;
  after_area_m2: number;
  difference_m2: number;
  percentage_change: number;
  primary_index: string;
  confidence: number;
  evidence_regions: EvidenceRegion[];
  metrics: Metric[];
  metadata?: Record<string, unknown>;
}

export interface AnalysisResult {
  status: AnalysisStatus;
  session_id: string;
  answer: string;
  confidence: number;
  confidence_available?: boolean;
  metrics: Metric[];
  evidence: EvidenceRegion[];
  trace: TraceStep[];
  mode: DataMode;
  demonstration_data?: boolean;
  vqa?: SingleImageVQAResult | null;
  caption?: SingleImageCaptionResult | null;
  bi_temporal_change?: BiTemporalChangeResult | null;
  cross_modal?: CrossModalOpticalSARResult | null;
  building_detection?: BuildingDetectionResult | null;
  building_temporal_change?: BuildingTemporalMatchResult | null;
  surface_area_change?: SurfaceAreaChangeResult | null;
}

export interface ApiResponse<T> {
  success: boolean;
  data: T;
}

export interface SubmitQueryData {
  session_id: string;
  result: AnalysisResult;
}

export interface AnalysisHistoryItem {
  session_id: string;
  created_at: string;
  query: string;
  intent?: string | null;
  mode?: string | null;
  status: string;
  summary_answer?: string | null;
  confidence?: number | null;
  metrics_summary?: Record<string, unknown> | null;
  input_summary?: Record<string, unknown> | null;
}

export interface AnalysisHistoryResponse {
  items: AnalysisHistoryItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface ErrorResponse {
  success: false;
  error: { code: string; message: string; user_message?: string | null; field?: string | null };
}

export class ApiError extends Error {
  code: string;
  userMessage: string;

  constructor(code: string, message: string, userMessage?: string | null) {
    super(message);
    this.name = "ApiError";
    this.code = code;
    this.userMessage = userMessage ?? message;
  }
}

File: backend/app/schemas/domain.py
from __future__ import annotations

from datetime import date, datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.schemas.vqa import SingleImageCaptionResult, SingleImageVQAResult
from app.schemas.bi_temporal_change import BiTemporalChangeResult
from app.schemas.imagery_policy import ImageryPolicyReport
from app.schemas.common import EvidenceRegion, GeoJSONGeometry, Metric

class SensorType(str, Enum):
    SENTINEL_2 = "sentinel-2"
    SENTINEL_1 = "sentinel-1"


class DataMode(str, Enum):
    DEVELOPMENT = "development"
    EARTH_ENGINE = "earth_engine"


class AnalysisStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class TraceStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"




class AOI(BaseModel):
    """Area of interest as GeoJSON geometry with optional metadata."""

    geometry: GeoJSONGeometry
    name: str | None = None
    area_km2: float | None = Field(default=None, ge=0)

    @model_validator(mode="after")
    def validate_polygon(self) -> AOI:
        if self.geometry.type not in ("Polygon", "MultiPolygon"):
            raise ValueError("AOI geometry must be Polygon or MultiPolygon")
        return self


class ImageryPreferences(BaseModel):
    cloud_cover_max: float = Field(default=30.0, ge=0, le=100)
    prefer_least_cloud: bool = True


class ImageryRequest(BaseModel):
    aoi: AOI
    start_date: date
    end_date: date
    sensor: SensorType = SensorType.SENTINEL_2
    preferences: ImageryPreferences = Field(default_factory=ImageryPreferences)
    demo_mode: bool = Field(
        default=False,
        description="When true, force deterministic demonstration imagery (no Earth Engine).",
    )

    @model_validator(mode="after")
    def validate_dates(self) -> ImageryRequest:
        if self.end_date < self.start_date:
            raise ValueError("end_date must be on or after start_date")
        return self


class SpatialMetadata(BaseModel):
    crs: str = "EPSG:4326"
    bbox: list[float] = Field(description="[min_lon, min_lat, max_lon, max_lat]")
    resolution_m: float | None = None


class ImageryScene(BaseModel):
    scene_id: str
    acquisition_date: date
    cloud_cover_percent: float | None = None
    preview_url: str | None = None
    platform_id: str | None = Field(
        default=None,
        description="Provider-native image identifier for downstream analysis (e.g. EE asset path)",
    )
    metadata: dict[str, Any] = Field(
        default_factory=dict,
        description="Provider-specific scene metadata (bands, tile, path/row, etc.)",
    )


class ImageryResult(BaseModel):
    source: str
    mode: DataMode
    sensor: SensorType
    scenes: list[ImageryScene]
    spatial: SpatialMetadata
    collection_id: str | None = Field(
        default=None,
        description="Source image collection identifier (e.g. COPERNICUS/S2_SR_HARMONIZED)",
    )
    provider_metadata: dict[str, Any] = Field(
        default_factory=dict,
        description="Provider-level metadata (selection policy, project, etc.)",
    )
    message: str | None = Field(
        default=None,
        description="Human-readable note, e.g. development adapter disclaimer",
    )


class QueryRequest(BaseModel):
    query: str = Field(min_length=3, max_length=2000)
    aoi: AOI | None = None
    earlier_date: date | None = None
    later_date: date | None = None
    sensor: SensorType = SensorType.SENTINEL_2
    preferences: ImageryPreferences = Field(default_factory=ImageryPreferences)
    image_id: str | None = Field(
        default=None,
        description="Uploaded image id for single-image VQA or scene-caption mode.",
    )
    earlier_image_id: str | None = Field(
        default=None,
        description="Earlier image id for uploaded bi-temporal change analysis.",
    )
    later_image_id: str | None = Field(
        default=None,
        description="Later image id for uploaded bi-temporal change analysis.",
    )
    optical_image_id: str | None = Field(
        default=None,
        description="Optical/multispectral image id for cross-modal analysis.",
    )
    sar_image_id: str | None = Field(
        default=None,
        description="SAR image id for cross-modal analysis.",
    )
    demo_mode: bool = Field(
        default=False,
        description="Use deterministic demonstration data instead of live Earth Engine catalog.",
    )

    @property
    def is_cross_modal_upload(self) -> bool:
        return self.optical_image_id is not None and self.sar_image_id is not None

    @property
    def is_bi_temporal_upload(self) -> bool:
        return (
            self.earlier_image_id is not None
            and self.later_image_id is not None
            and not self.is_cross_modal_upload
        )

    @property
    def is_single_image_vqa(self) -> bool:
        return (
            self.image_id is not None
            and not self.is_bi_temporal_upload
            and not self.is_cross_modal_upload
        )

    @model_validator(mode="after")
    def validate_query_dates(self) -> QueryRequest:
        if self.is_cross_modal_upload:
            if self.image_id or self.earlier_image_id or self.later_image_id:
                raise ValueError(
                    "cross-modal mode requires only optical_image_id and sar_image_id"
                )
            if self.aoi is not None or self.earlier_date is not None or self.later_date is not None:
                raise ValueError("aoi and catalog dates are not used for cross-modal upload analysis")
            return self
        if self.is_bi_temporal_upload:
            if self.image_id is not None:
                raise ValueError("image_id cannot be combined with earlier_image_id/later_image_id")
            if self.aoi is not None or self.earlier_date is not None or self.later_date is not None:
                raise ValueError(
                    "aoi and catalog dates are not used for uploaded bi-temporal change analysis"
                )
            return self
        if self.is_single_image_vqa:
            if self.earlier_image_id or self.later_image_id or self.optical_image_id or self.sar_image_id:
                raise ValueError("single-image mode requires image_id only")
            return self
        if self.earlier_image_id or self.later_image_id:
            raise ValueError("earlier_image_id and later_image_id must be provided together")
        if self.optical_image_id or self.sar_image_id:
            raise ValueError("optical_image_id and sar_image_id must be provided together")
        if self.aoi is None or self.earlier_date is None or self.later_date is None:
            raise ValueError("aoi, earlier_date, and later_date are required for catalog queries")
        if self.later_date <= self.earlier_date:
            raise ValueError("later_date must be after earlier_date")
        return self




class TraceStep(BaseModel):
    id: str
    tool_name: str
    status: TraceStatus
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = None
    summary: str | None = None
    error: str | None = None
    metadata: dict[str, object] | None = None


class EvidenceOverlay(BaseModel):
    model_config = ConfigDict(extra="forbid")

    overlay_id: str
    image_id: str
    supported_layers: list[str] = Field(default_factory=list)
    preview_url: str | None = None
    data_uri: str | None = None
    width: int = Field(ge=1)
    height: int = Field(ge=1)
    crs: str | None = None
    bounds: list[float] | None = None
    statistics: dict[str, Any] = Field(default_factory=dict)


class AnalysisResult(BaseModel):
    status: AnalysisStatus
    session_id: str
    answer: str
    confidence: float = Field(ge=0, le=1)
    confidence_available: bool = True
    metrics: list[Metric] = Field(default_factory=list)
    evidence: list[EvidenceRegion] = Field(default_factory=list)
    trace: list[TraceStep] = Field(default_factory=list)
    mode: DataMode = DataMode.DEVELOPMENT
    demonstration_data: bool = Field(
        default=False,
        description="True when the user explicitly requested demo_mode on a catalog query.",
    )
    vqa: SingleImageVQAResult | None = None
    caption: SingleImageCaptionResult | None = None
    bi_temporal_change: BiTemporalChangeResult | None = None
    cross_modal: "CrossModalOpticalSARResult | None" = None
    imagery_policy: ImageryPolicyReport | None = None
    building_detection: "BuildingDetectionResult | None" = None
    building_temporal_change: "BuildingTemporalMatchResult | None" = None
    surface_area_change: "SurfaceAreaChangeResult | None" = None
    evidence_overlay: EvidenceOverlay | None = None
    domain: str | None = None


class ChangeDetectionInput(BaseModel):
    aoi: AOI
    earlier_date: date
    later_date: date
    imagery: ImageryResult
    query_hint: str | None = Field(
        default=None,
        description="Optional natural-language hint for index selection on uploaded pairs.",
    )
    change_domain: str | None = Field(
        default=None,
        description="Phase 5B/7 change domain for catalog index routing (earth_engine path).",
    )


class ChangeDetectionOutput(BaseModel):
    regions: list[EvidenceRegion]
    raw_detection_count: int
    detector: str
    mode: DataMode
    detector_metadata: dict[str, Any] = Field(
        default_factory=dict,
        description="Detector-level metadata (method, threshold, scene ids, etc.)",
    )


class FetchImageryInput(BaseModel):
    request: ImageryRequest


class FetchImageryOutput(BaseModel):
    result: ImageryResult


class GenerateEvidenceInput(BaseModel):
    query: str
    imagery: ImageryResult
    fused_regions: list[EvidenceRegion]
    fusion_metadata: dict[str, Any] = Field(default_factory=dict)


class FuseEvidenceInput(BaseModel):
    cva_detections: ChangeDetectionOutput
    semantic: SemanticAnalysisOutput | None = None
    sar_detections: ChangeDetectionOutput | None = None


class FuseEvidenceOutput(BaseModel):
    regions: list[EvidenceRegion]
    fusion_metadata: dict[str, Any] = Field(default_factory=dict)


class DetectSARChangeInput(BaseModel):
    aoi: AOI
    earlier_date: date
    later_date: date
    preferences: ImageryPreferences = Field(default_factory=ImageryPreferences)


class GenerateEvidenceOutput(BaseModel):
    regions: list[EvidenceRegion]
    metrics: list[Metric]
    confidence: float


class SemanticClaim(BaseModel):
    """Structured semantic claim produced by a SemanticAnalyzer."""

    claim_type: str
    region_id: str
    confidence: float = Field(ge=0, le=1)
    metrics: list[Metric] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class SemanticAnalysisInput(BaseModel):
    aoi: AOI
    earlier_date: date
    later_date: date
    imagery: ImageryResult
    change_regions: list[EvidenceRegion]
    analysis_profile: str


class SemanticAnalysisOutput(BaseModel):
    regions: list[EvidenceRegion]
    claims: list[SemanticClaim] = Field(default_factory=list)
    analyzer: str
    mode: DataMode
    analyzer_metadata: dict[str, Any] = Field(default_factory=dict)


from app.schemas.cross_modal import CrossModalOpticalSARResult  # noqa: E402
from app.schemas.building_analysis import BuildingDetectionResult, BuildingTemporalMatchResult  # noqa: E402
from app.schemas.surface_change import SurfaceAreaChangeResult  # noqa: E402

AnalysisResult.model_rebuild()
